### ISSUE TITLE
Input Lag

### ISSUE THREAD

**Issue by jupham** [chrippa/ds4drv#42]
I didn't want to put this as a bug but I am trying to get more info on this issue. I got ds4drv working in steamos but I'm noticing a little bit of input lag. It isn't a ton but it does affect usability when compared with having the controller hooked up through usb. Is this noticeable for others?

I could only get it to work when not in hidraw mode. Would having hidraw mode working improve upon this?

**Comment by atomnuker** [comment]
I should probably mention that (with BT-400, chipset is BCM20702) there's only a low report rate under Linux. Under Windows there's a constant 1ms/update so it's most likely a driver issue. Maybe there's a difference between how Linux and Windows initialize the device.
I'll take a look at the kernel code, but at this point it's almost certain that all low report rate problems are from the drivers. Since some users report that with other devices there's no problem, BlueZ is probably not the one to blame.

**Comment by markubiak** [comment]
Yup.  I reported this in several places but everyone keeps putting the blame at different programs and drivers.

**Comment by atomnuker** [comment]
The place where CSR8510 (the !only! listed chipset as "working fine") and BCM*/Intel mostly diverge after initialization is by setting different HCI quirk bits.
Does anyone know other fine? Other than the ones listed in the wiki?
Moreover, is CSR8510 really fine and reporting at 1000Hz? Is there anyone with that hardware who can verify that they get more than 22 updates/sec? I'm really doubting the "works" report on that one because it's a chipset known to not be very standardized across revisions and has a few quirks of its own.

HCI_QUIRK_SIMULTANEOUS_DISCOVERY seems to be the one flipped, I might try recompiling my btbcm module with that one set tommorow. If there isn't a single device which is verified to work then the problem's in the HCI layer (event handling maybe) which sits in between hid-sony.c and btusb.c.

**Comment by atomnuker** [comment]
I can definitely, 100% confirm that it is not the fault of the Bluetooth device.
Tested both the "Good and working" Telewell (It uses a generic CSR8510, I got a similar one from China) and the Broadcom BCM20702A (used by Asus, etc.). Removed the Telewell POS from the wiki to avoid having people fooled by it.
Whatever's happening is either BlueZ's problem (probably not since my bluetooth mouse updates at 125 Hz) or the hid-sony driver's fault.
Probably the driver isn't sending a command it should often enough. It's a fairly small driver, at least the DS4 part so it shouldn't be that difficult to track down.
DS4 will print out limited information without any sensors if 0x02 isn't received but maybe it needs to have that sent out more often.

**Comment by atomnuker** [comment]
Found it.
It was a one line fix. A bloody one line fix during initialization. Now I'm getting 1000 reports a second.

I'll send out a patch out today and hopefully it'll get merged before 4.2 gets released.

**Comment by markubiak** [comment]
Excellent work!  I looked through the sonyhid driver but didn't know enough to know what was happening.  What was the issue, and where can I find the patch?

**Comment by atomnuker** [comment]
Here's the patch:
http://www.mail-archive.com/linux-kernel@vger.kernel.org/msg941718.html

The issue was a single letter.

**Comment by jupham** [comment]
Wow, great find. I'm not familiar with applying patches like this, any idea how to get into steamos (debian)? My whole reason for starting this was to get it working there.

**Comment by markubiak** [comment]
From my understanding you'll have to compile your own kernel and apply this patch to the hid-sony module.  It should work on any version of the kernel that has this module.

**Comment by markubiak** [comment]
I can 100% confirm that atomnuker's patch fixes the issue on Arch 4.1.3-1.  Loving playing games without a cord, thanks!

**Comment by Ape** [comment]
@markubiak, Did you patch the kernel yourself? Is there an AUR package available?

**Comment by markubiak** [comment]
Place hid-sony.ko.gz in /usr/lib/modules/4.1.3-1-ARCH/kernel/drivers/hid.  It's in a zip.  If there's demand I'll look on how to make a AUR package for it.

[Download](https://uofi.box.com/shared/static/o1ma9zx2tfh8qvniyaxnag6bf8i662yt.zip)

**Comment by jacobmischka** [comment]
Do you happen to have that file still? The link seems to be broken.

**Comment by Ape** [comment]
I just checked that this has been merged to Linux 4.3. (Yes, you still need a patch before it's released.)

**Comment by markubiak** [comment]
I actually permanently deleted the patch, as it was only compiled for an outdated kernel.  There are instructions on the arch wiki for compiling kernel modules, the file is in drivers/hid/ and called hid-sony.  The previously linked patch request has information on how to modify the file.  If you need any help let me know.

**Comment by jacobmischka** [comment]
Yeah I know, I'm just really lazy. Thanks though.

**Comment by colincrossland** [comment]
Earlier it was mentioned that this was patched in 4.3 but I'm still having the same problem after update. Anyone else?

**Comment by ShadwDrgn** [comment]
I suspect I'm having this issue as well. I'm on kernel 4.4 and using git cloned today for ds4drv. I'm seeing a lot of  signal strength low messages and my controllers seem fine sometimes but others seem to have up to a second or more of input lag. It's extremely erratic when the lag happens. I've tried hidraw mode with updated bluez and without hidraw. Just want my controllers to work with retroarch and steam :~(

**Comment by brewster1134** [comment]
is this still an open issue? i have a fresh build of retropie/ds4drv and have very low signal strength (between 2-20).  i have 2 controllers and was not able to connect with hidraw.

i am on ds4drv v 0.5.1 so it seems like this fix would have already been applied, but im not sure where this patch is actually applied to check. i dont see a `hid.sony` file in this repo.

**Comment by jacobmischka** [comment]
Yeah, it's still a problem. I've basically just given up with bluetooth on ds4, I just use a cable.

**Comment by vegham** [comment]
Still issue here as well. I have the ASUS BT400 BCM20702A0 device.

**Comment by sampiforce** [comment]
I am currently deep into this issue as well.  New retropie, new bluez, new ds4drv.  I am running a pi zero with a redbearIoT pHAT as the bluetooth module.
I cant connect in hidraw mode either, I am not getting more than 20 reports/sec and I KNOW this guy was on to something with this fix.  WHERE DO WE MAKE THIS CHANGE???  It is killing me thinking that we are so close.
